### ISSUE TITLE
Server communicate with LED controller

### ISSUE THREAD

**Issue by KevinGage** [antimeme/rotpi#7]
Add logic to server to process requests to turn on/off LED lights and send them to controller.

**Comment by KevinGage** [comment]
@jorgelr617  @antimeme  @mikegomezg  @JohnnyRobotic  @amurray22 

I've got a thought about how this could work.  I want to see what others think.  We'll need to get this worked out before we can write the logic in our clients.

So we know the client will build a command and it will want to send it to the server.  So the client needs a mechanism to send to the server and the server needs a mechanism to receive the command and process it.  I think we agree that the client can use AJAX to send the command.

I think that we could have a url on the server that accepts commands from the client.  Lets call it 'blinklights' just for now.  The blinklights page could check for a http GET paremeter again for arguments sake lets call it 'command'.  Because of the nature of URL's it is hard to send a bunch of commands cleanly so I propose using basse64 encoding to encode the command before it's sent, and decode it on the server when received.  This isn't as scary as it sounds and a simple example is below.

So here is an example of the workflow in this scenario
1. A person visits site and does some things to decide what lights will flash.  Then they click the go button.
2. The client page translates this into a command like this '{delay:0, 17:true}'.  Let's not worry too much about the details of this step until we agree on the rest.
3. The client page encodes the command like this 'var encoded = btoa({delay:0, 17:true});'  This will make a variable called encoded will will contain the encoded value.  That's it!  the encoded value will look like this 'e2RlbGF5OjAsIDE3OnRydWV9'. 
4. The client page sends the request to the server using AJAX like this 
   http://{server ip address}/blinklights?command=e2RlbGF5OjAsIDE3OnRydWV9
5. The server receives the address and decodes the command like this 
   'var decoded = atob(e2RlbGF5OjAsIDE3OnRydWV9)'
6. Now the server can process the command in any way that it wants.

btoa and atob are built in javascript functions to do the encoding and decoding.
http://www.w3schools.com/jsref/met_win_btoa.asp

Sorry about the length I'm just trying to write this as clearly as possible.

**Comment by antimeme** [comment]
On Wed, Jul 15, 2015 at 9:21 AM, KevinGage notifications@github.com wrote:

> The blinklights page could check for a http GET paremeter again for
> arguments sake lets call it 'command'.
> 
> REST suggests that GET is only used for operations without side effects.
> If we're changing the state of an LED I'd suggest we use POST for that.
> This also complicates some forms of attack that involve tricking a user
> into clicking a link and avoids complications with caching proxies.
> 
> Because of the nature of URL's it is hard to send a bunch of commands
> cleanly so I propose using basse64 encoding to encode the command before
> it's sent, and decode it on the server when received. This isn't as scary
> as it sounds and a simple example is below.
> 
> Base64 is easy enough, but I'm curious about why we want to use the URL to
> encode commands.  Why not just use POST we can send JSON as the body of the
> HTTP request?  This seems like exactly what the HTTP request body is meant
> to be used for.  Here's a jQuery POST example I found on the web:

var postData={x:1,y:2};
$.post("http://AJAX_POST_URL",
    postData,
    function(data, textStatus, jqXHR)
    {
        //data - JSON object from server.
    },"json").fail(function(jqXHR, textStatus, errorThrown) {});

I think that's a bit simpler than encoding things, don't you?

**Comment by KevinGage** [comment]
The encoding wouldn't be needed with POST.  For GET it would be needed because the URL http://whatever.com/lights?command={stuff,more stuff} is going to get url encoded and break.  That wouldn't be a problem with POST.
I think you are right and using an HTTP POST request simplifies things and is a good idea.

**Comment by jorgelr617** [comment]
@jorgelr617 @antimeme @mikegomezg @JohnnyRobotic @amurray22

Kevin, this task is "Server communicates with LED controller". I'm assuming that your client question is another task. So I will answer the client task question first. So no, we don't have to use AJAX to have the client communicate with the sever. I think this can be done beautifully in-real time with WebSockets or the Socket.IO framework. I think it is a better architectural solution.  I don't mind volunteering for this task. The message can then be easily encoded/decoded in JSON.

About "Server communicate with LED controller" task, you have some interesting challenges to handle. I'll mention one. How do you process multiple clients sending messages to the server that conflict turning on/off the LEDs at the same time? Etc. To me, that's the main essence of this task.

**Comment by KevinGage** [comment]
@jorgelr617 @antimeme @mikegomezg @JohnnyRobotic @amurray22
Yes there will have to be some design planning and discussions to decide how we want to handle the distributed nature of the clients. Since the controller will use callbacks we could use some queueing mechanism to wait for one task to complete before starting the next.  Or we could fun wild west and just let instructions all flow together.
On another note if anyone is sick of spam from these conversations let us know and we can stop adding you to every comment.

**Comment by jorgelr617** [comment]
@antimeme @mikegomezg @JohnnyRobotic @amurray22 @KevinGage 

Kevin, what do you think about my other recommendation?

"So no, we don't have to use AJAX to have the client communicate with the sever. I think this can be done beautifully in-real time with WebSockets or the Socket.IO framework. I think it is a better architectural solution. I don't mind volunteering for this task. The message can then be easily encoded/decoded in JSON."

**Comment by KevinGage** [comment]
@antimeme @mikegomezg @JohnnyRobotic @amurray22 @KevinGage
I think it could be done either way.  I think web sockets would make it possible to get feedback on what others are submitting to the server which is cool. I haven't worked with them yet, only read about them.

**Comment by antimeme** [comment]
WebSockets are a great technology but they feel like overkill here.  All we
need is a few messages back and forth to the server.  We would need
WebSockets for voice or video data, for example.  However, AJAX is dead
simple to implement -- especially with jQuery.  Why don't we start with
AJAX for now and think about an application for WebSockets later?

**Comment by jorgelr617** [comment]
@antimeme @mikegomezg @JohnnyRobotic @amurray22 @KevinGage

Jeff, the Server needs to be able to send to all the clients the commands entered by the other clients, which is like an asynchronous notification. How can you do that with AJAX? I think that it can only be done with WebSockets.

**Comment by antimeme** [comment]
What you're describing is easy to do using AJAX.  I think you mean that
AJAX can't do server push.  Is that right?  If so that's technically true
but it hasn't stopped people from getting close enough using various
polling strategies since the 1990's.  It's just a matter of having the
browser code call "setTimeout(fn)" and make an AJAX call in there.  Does
that make sense?

Yes, WebSockets is better than this.  As I said, it's a great technology,
but it's too complex for what we're trying to do.  It's much better to
start with something simple and build up than to try creating some kind of
enterprise class system from scratch.

**Comment by jorgelr617** [comment]
@antimeme @mikegomezg @JohnnyRobotic @amurray22 @KevinGage

Yes, I mean a server push- such as notifications. I don't agree that it is a complicated solution for what we are trying to do. You only have to handle a few events, like connect, message, error, etc. Not really different in programming complexity than AJAX. I think it fits nicely by allowing clients and servers to communicate with each other in real-time. I think polling is clunky and chatty. Most folks probably won't be able to make this week's MeetUp, but I can have a WebSockets demo ready for the next MeetUp and we can talk more. 

I'm also hoping at the next MeetUp that we can discuss the vision of the project, design and future tasks. We need to think of ways to allow everyone to contribute.

**Comment by KevinGage** [comment]
@antimeme @mikegomezg @JohnnyRobotic @amurray22 @jorgelr617 
We could always provide both options.  One URI could accept AJAX requests, while another does whatever is necessary for websockets.
Then we could use this as a learning device to learn the pros and cons of each method.
Just a thought.

**Comment by jorgelr617** [comment]
Great idea Kevin.I wonder if there is anything that can stops us from doing both?

**Comment by KevinGage** [comment]
@antimeme  @jorgelr617  @mikegomezg  @JohnnyRobotic  @amurray22 
I can start working on the AjAX through HTTP POST approach.  I think I would first just set it up to accept commands and issue them to the LED controller.  Then once it was working we could add in logic to decide how to deal with multiple requests.  Like a command queue or something.
Then Jorge maybe you could work on the websockets method since you have some experience in that area?
Does that sound ok?

**Comment by jorgelr617** [comment]
@antimeme @jorgelr617 @mikegomezg @JohnnyRobotic @amurray22

Kevin, I don't know enough to assess whether you can use both AJAX and WebSockets in the same server. Why not reach a compromise? I definitely want to make  the "way ahead" decision as a group in our next MeetUp. In the mean time, why don't you implement the AJAX solution privately in your local computer and demo it next week. (In other words, don't check in your changes). I'll do the same for WebSockets. Also, have you considered working on your client?

**Comment by KevinGage** [comment]
@antimeme @JohnnyRobotic @jorgelr617  @mikegomezg  @amurray22 
That sounds good.  I'll make a branch for ajax and you can branch for websockets.  Then later we can try to merge them and see how things go.
I do have an idea for my client.  I just wanted to be able to test my client with the server so the AJAX or websocket solution would need to be in place first.  I'll try and have both done for next saturday.  At least early versions of them.
Sounds good?

**Comment by jorgelr617** [comment]
Yes, sounds good.

**Comment by mikegomezg** [comment]
You guys already arrived at the solution I think makes sense, but I just wanted to point something out.

Which technology we use to implement the communication between client and server is important because it determines what we end up learning more about.

Learn is the key word here. If this were a real app, then yeah we should agree on the technology first. But duplicating functionality with different tech, though it may seem inefficient, creates interesting contrasts as noted above.

I'm excited to see what Kevin and Jorge come up with! I'm working on my AngularJS presentation by the way, so I wont be contributing much right now. But I'll be following along.

**Comment by KevinGage** [comment]
I have the AJAX calls working on the server and I have a very simple client.  I'll work on polishing the server and making a better client for saturdays meetup.

**Comment by jorgelr617** [comment]
@antimeme @JohnnyRobotic @jorgelr617 @mikegomezg @amurray22 @TheTallOne0606

Mike, I don't have access to the Raspberry Pi and the circuit board, so I won't be running my demo in "real-mode". (I would need those things and time to debug it.) I'll be showing how easy it is to send/receive messages with WebSockets or Socket.io by just having the Server printing the messages to the console. I agree with you that learning is the key word here. I think we can explore many technologies with this project.

Kevin, this leads me to think about another idea. We need some kind of simulation/off-line mode so that when your LED functions are called without the circuit board, they print to the console the command and eventually simulate the LEDs (with graphics) in the status page that Jeff is working on. By the way Kevin, what kind of job do you have? :-) How do you find so much free time to work on the project? :-)

Jeff, at this week's MeetUp, I hope we can walk through the server code to help us understand the logic.  I also have some GitHub questions that you probably can answer.

**Comment by KevinGage** [comment]
@antimeme @JohnnyRobotic @jorgelr617 @mikegomezg @amurray22 @TheTallOne0606
I am a network engineer and manager for a small IT managed services company.  We essentially are the IT department for companies that don't have their own IT staff.  I don't typically do any coding as part of my job and just enjoy it as a hobby.
If everything is running smoothly at the office I do get some time to spend researching or coding while at the office.  Otherwise I'm just doing everything on my off time.  I don't have and children so in my free time I can do whatever I want.
Yesterday I did about an hour of work in the morning while drinking my coffee, then and 1.5-2 hours when I got home at night while watching tv.
I agree we do need a simulation mode since not everyone will have a pi.  I want to get my client wrapped up before saturday and have some work to do to make the server to led controller better so I probably won't have time to look at a simulation mode at the moment.
I'm really looking forward to the websockets demo.  I know about web sockets but have never worked with them before.

**Comment by antimeme** [comment]
I have a bunch of first generation RaspberryPi devices going spare.  I
bought them a few years ago for a project that never came together.  I
believe I suggested this in the previous meetup but again I don't mind
lending these out or even giving them away.  So, if you want one to play
with at home just let me know.  As an alternative, you can also purchase a
kit (as I believe Kevin has done) at http://adafruit.com/ for about $100.
Either way.

I'm working on Issue #5 but I can make a simulation mode part of that.
This will be helpful even for people who do have a Pi since sometimes it's
more pleasant to tinker and test without unpacking a pile of equipment.

**Comment by KevinGage** [comment]
One other note.  The onoff package doesn't seem to be windows compatible.  So it looks like you need to be running it on linux for testing even if you don't intend on doing anything with the lights.  Maybe that can be circumvented by the simulation mode.

**Comment by antimeme** [comment]
That makes sense.  I'll probably remove the onoff from the repository at
some point.  We'll make the installation process include an "npm install"
somehow.  That can be how we know whether to use the simulator or not.

**Comment by KevinGage** [comment]
@antimeme  @jorgelr617  @mikegomezg  @amurray22  @JohnnyRobotic 
fun with a multicolored led
https://drive.google.com/open?id=0B6JmKqU3aVSZczdZejA0OTh3SHc

**Comment by antimeme** [comment]
Nice!

-- Jeff
